Make Weather Page more useful for Network Monitoring [2/2]

raleighlittles opened the pull request:
This PR addresses the remaining tasks from [the story](https://www.pivotaltracker.com/n/projects/1314272/stories/156576199) that were not addressed in the 1st pull request, the main features being:

1. Now displays number of sites opened/closed/unknown in the sidebar along with each's site basic status (Y, N, or '?' for unknown).

2. Adjusted the labeling for closure thresholds to distinguish between minimum value thresholds or maximum value thresholds.

3. Removed `node-sass` package, which according to npm, had the vulnerable package 'hoek' in its dependency tree. 

4. Now shows the times the site was closed on its "Ok to Open" plot, and hovering over the closed section shows a tool-tip which tells you why the site was closed.

5. Added a 14-day option.

markBowman commented:
The client fails to build for me:

ERROR in /node_modules/css-loader!/node_modules/vue-loader/lib/style-compiler?{"vue":true,"id":"data-v-3a24dfec","scoped":false,"hasInlineConfig":false}!/node_modules/sass-loader/lib/loader.js!/node_modules/vue-loader/lib/selector.js?type=styles&index=0!/src/App.vue
Module build failed: Error: Cannot find module 'node-sass'

raleighlittles commented:
@markBowman My apologies, it worked when I tested it locally (using NPM) and I assumed that meant it would be okay when built with Docker. 

I tested it just now with docker and was able to build and run successfully:
```
[rlittles@rlittles-linux1.lco.gtn client]$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED              STATUS              PORTS                NAMES
265509512d6f        weatherclient       "node /srv/weather..."   About a minute ago   Up About a minute   0.0.0.0:80->80/tcp   serene_lumiere
c71907a58ac0        weatherserver       "go-wrapper run"         About a minute ago   Up About a minute   80/tcp               tender_dijkstra
```

The weird issue that I'm encountering now is that Docker is still showing that we have a security vulnerability in our dependency tree, even though NPM shows that I'm using the latest version of the package that's that vulnerability. 
Running ```$ npm ls hoek```gives me:
```[rlittles@rlittles-linux1.lco.gtn client]$ npm ls hoek
lcoweather@1.0.0 /home/rlittles/git_projects/weather/client
└─┬ hawk@7.0.7
  └── hoek@5.0.3 
```
but during the Docker image build step, it still gives the vulnerability error:

```
npm WARN notice [SECURITY] hoek has 1 moderate vulnerability. Go here for more details: https://nodesecurity.io/advisories?search=hoek&version=2.16.3 - Run `npm i npm@latest -g` to upgrade your npm version, and then `npm audit` to get more info.
```
You can see that earlier, NPM is reporting that its only the safe version of hoek (5.0.3) is installed, so there appears to be some conflicting info here.

markBowman commented:
Air Temperature threshold is a minimum, so should be green in the new way of displaying thresholds

markBowman commented:
TFN open state is showing as unknown but it is open. What causes this?

raleighlittles commented:
@Fingel @markBowman To add on to the security vulnerability issue, it looks like we've reached an impasse: the ```node-sass``` package (https://www.npmjs.com/package/node-sass) that we use is requiring an older version of the ```request``` package, which has the unsafe package in its dependecy tree.

```
[rlittles@rlittles-linux1.lco.gtn client]$ npm ls hoek
lcoweather@1.0.0 /home/rlittles/git_projects/weather/client
├─┬ hawk@7.0.7
│ ├─┬ boom@7.2.0
│ │ └── hoek@5.0.3  deduped
│ ├── hoek@5.0.3 
│ └─┬ sntp@3.0.1
│   ├─┬ bounce@1.2.0
│   │ ├─┬ boom@7.2.0
│   │ │ └── hoek@5.0.3  deduped
│   │ └── hoek@5.0.3 
│   └── hoek@5.0.3  deduped
└─┬ node-sass@4.9.0
  └─┬ request@2.79.0
    └─┬ hawk@3.1.3
      ├─┬ boom@2.10.1
      │ └── hoek@2.16.3  deduped
      ├── hoek@2.16.3 
      └─┬ sntp@1.0.9
        └── hoek@2.16.3  deduped
```

Someone opened an issue on the node-sass project a while ago (see: https://github.com/sass/node-sass/issues/2355) but its still open. I tried reverting to an older version of node-sass (4.7.1) but this had an issue building on my system (see: https://github.com/nodejs/node-gyp/issues/1217). Not quite sure what to do here.

Fingel commented on client/src/App.vue:
```diff
@@ -1,14 +1,14 @@
   <div id="app">
     <section class="section">
-      <div class="container">
+      <div class="container is-fluid is-marginless">
```
can we remove is-marginless? The page is much easier to read with a bit of a margin.

Fingel commented on client/src/components/Site.vue:
```diff
@@ -18,59 +18,59 @@
           <div>
             <p class="heading">Pressure mbar</p>
-            <p class="title">{{ datums['Weather Barometric Pressure Value'].data | latestVal }}</p>
+            <p class="title">{{ this.$options.filters.latestResult(datums['Weather Barometric Pressure Value'].data, 'Value') }}</p>
```
if this is a filter, it should be used as a filter, like this: `{{ datums['Weather Barometric Pressure Value'] | latestResult }}`

raleighlittles replied:
Duly noted. VueJS shows that the two syntaxes are equivalent so I didn't think it mattered which was used; but its fixed in the latest commit.

Fingel commented on client/src/components/TimeChart.vue:
```diff
@@ -62,17 +62,26 @@ export default {
+              {
+                enabled: true,
+                position: 'left',
+                content: (this.limit_direction === 'min') ? 'minimum' : '',
```
This label is super distracting, I think it would be a lot better as a legend.

raleighlittles replied:
I don't quite think this is supported by the plugin that we use. ChartJS obviously has legends,(https://www.chartjs.org/docs/latest/configuration/legend.html) but I don't see any support for legends in the ChartJS annotation plugin. What I could maybe try doing is only showing the annotation label if you hover over the chart?

Fingel commented on client/src/components/WeatherNav.vue:
```diff
@@ -44,22 +56,106 @@
+            }
+            else {
+              var last_val = resp[resp.length - 1].ValueString;
+              var last_letter = this.$options.filters.statusMsgToLetter(last_val);
```
Filters are only meant to be used in templates. If this is just a helper method, no need to place it in filters. Use a computed or normal method instead.

raleighlittles replied:
Okay, this code was moved to the methods() section.

raleighlittles commented:
@Fingel @markBowman The changes requested are up, but with regards to the label as a legend, I couldn't quite find a way to display the label as a legend. My approach now was to hide the label, and instead let the user click on the threshold line to toggle the label (minimum or maximum). I originally wanted to use the mouseover/mouseleave events (so you when you hover over the plot it shows up, and when you leave it goes away) but the plugin we used appears to have issues with events (https://github.com/chartjs/chartjs-plugin-annotation/issues/88) so I couldn't quite get that working. 

I also added some text below the first plot to tell the user about the functionality -- see: 

![screenshot from 2018-05-29 17-50-08](https://user-images.githubusercontent.com/2137201/40748285-c8a97894-6414-11e8-801e-df2c51f7a9ca.png)

Let me know what you think.

Fingel commented:
Looks good!

markBowman commented:
The legend to the Sky transparency minimum green line has a value of 0 even though the line is at 25%

markBowman commented:
OPINION: I would lose the Minimum/Maximum legend from the left side panel because it is not obvious it applies to the plots. You could put it at the very bottom of the main panel under all the plots.

markBowman commented:
We're getting into the real nitty-gritty of things you weren't asked to do but....
* "current values" should be bold (maybe) to make it clear it is a section label
* The time period title (e.g. "LAST 24 Hours") is also hard to pick out. Possibly add some space or line separation from the current values. You could also move the Time range picker to here rather than duplicating the information.

markBowman commented:
If I increase the size of my browser window the graph scale to take advantage but if I shrink it again they do not get smaller so the right side is out of view.

markBowman commented:
Sorry, didn't mean to hit close.

The green and yellow widgets with number of open and closed sites are clickable buttons with no obvious behavior.

raleighlittles commented:
@markBowman @Fingel Thanks for the feedback. The latest commit has the following changes.

* Increased size of "Current Values" heading
* Increased size of "Last <time range>" heading 
* Moved legend from sidebar to underneath all plots
* Changed buttons to [tags](https://bulma.io/documentation/elements/tag/) so that they're not clickable but still look the same
* Removed the minimum/maximum value display from the plot annotation -- I'm not sure why the values weren't working for the Sky Transparency closure, since they worked fine for all the other plots. This wasn't defined in the scope of the story anyways, and I don't think being able to tell the *exact* value of the threshold is a huge deal, as you can easily discern what the threshold is with relatively good accuracy (i.e. do we really need to tell if the threshold is at 25 and not 22.5, etc.)

Regarding the issue about page resizing, you have to refresh the page if you change the size of the window in the x direction -- the existing version of the LCO Weather page has similar behaviour. 

Below are screenshots of the changes.
![screenshot from 2018-05-31 13-13-01](https://user-images.githubusercontent.com/2137201/40805840-71e39f24-64d4-11e8-995f-ccec3e6fb5ee.png)
![screenshot from 2018-05-31 13-13-20](https://user-images.githubusercontent.com/2137201/40805841-7200987c-64d4-11e8-8aec-0f89e49ceaa7.png)

wrosing commented:
MUST we continue to call Sky Brightness that instead of Sky Darkness?

On Thu, May 31, 2018 at 8:16 PM, Raleigh Littles <notifications@github.com>
wrote:

> @markBowman <https://github.com/markBowman> @Fingel
> <https://github.com/Fingel> Thanks for the feedback. The latest commit
> has the following changes.
>
>    - Increased size of "Current Values" heading
>    - Increased size of "Last " heading
>    - Moved legend from sidebar to underneath all plots
>    - Changed buttons to tags
>    <https://bulma.io/documentation/elements/tag/> so that they're not
>    clickable but still look the same
>    - Removed the minimum/maximum value display from the plot annotation
>    -- I'm not sure why the values weren't working for the Sky Transparency
>    closure, since they worked fine for all the other plots. This wasn't
>    defined in the scope of the story anyways, and I don't think being able to
>    tell the *exact* value of the threshold is a huge deal, as you can
>    easily discern what the threshold is with relatively good accuracy (i.e. do
>    we really need to tell if the threshold is at 25 and not 22.5, etc.)
>
> Regarding the issue about page resizing, you have to refresh the page if
> you change the size of the window in the x direction -- the existing
> version of the LCO Weather page has similar behaviour.
>
> Below are screenshots of the changes.
> [image: screenshot from 2018-05-31 13-13-01]
> <https://user-images.githubusercontent.com/2137201/40805840-71e39f24-64d4-11e8-995f-ccec3e6fb5ee.png>
> [image: screenshot from 2018-05-31 13-13-20]
> <https://user-images.githubusercontent.com/2137201/40805841-7200987c-64d4-11e8-8aec-0f89e49ceaa7.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/LCOGT/weather/pull/6#issuecomment-393665818>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKedsAO9Hr53DKvWevn0fKTfOIbl6S4pks5t4E-xgaJpZM4ULZla>
> .
>



-- 
Clear skies,

Wayne Rosing
Founder
805 708 6901 cell
805 880 1603 office

Please note wrosing@lco.global is our preferred address.
